v2/tools/generator/internal/codegen/pipeline/state.go

```diff
@@ -51,10 +57,31 @@ func (s *State) WithConversionGraph(graph *storage.ConversionGraph) *State {
 		panic("may only set the conversion graph once")
 	}
 
-	return &State{
-		definitions:     s.definitions.Copy(),
-		conversionGraph: graph,
+	result := s.copy()
+	result.conversionGraph = graph
+	return result
+}
+
+// WithSeenStage records that the passed stage has been seen
+func (s *State) WithSeenStage(id string) *State {
+	result := s.copy()
```

Comment:
Do we need copy of the whole struct here? We can only get `stagesSeen` and `stagesExpected` into vars independently and do the manipulation?

Reply:
We used to have `stagesSeen` and `stagesExpected` in local variables; this is why we ended up with tests that weren't configured properly. Moving them into `State` guarantees they're carried forward correctly.

We return a copy because `State` is (effectively) an immutable type that should never be changed after instantiation; this allows (for example) comparision of pre- and post- stage values to see what changed.